src/main/scala/com/mozilla/telemetry/mozdata/MozData.scala

```diff
@@ -0,0 +1,335 @@
+    log("sql", Map("query"->Some(query)))
+    spark.sql(query)
+  }
+
+  /** Write partition to long term storage
+    *
+    * @param df DataFrame to write
+    * @param tableName table to write
+    * @param partition ordered list of key-value static partition identifiers,
+    *                  which must be absent from df. optional if specifying uri
+    * @param version optional specific version of table, defaults to "v1" for
+    *                new tables and the latest version for existing tables
+    * @param owner optional email that identifies non-global namespace
+    * @param uri optional non-standard location for this table
+    * @param metadataUpdateMethods optional methods to use to update metadata
+    *                               after writing partitioned global tables
+    * @param extraWriteConfig optional function to configure the DataFrameWriter
+    */
+  def writePartition(df: DataFrame, tableName: String,
+                     partition: List[(String,String)] = List(),
+                     version: Option[String] = None,
+                     owner: Option[String] = None, uri: Option[String] = None,
+                     extraWriteConfig: Function[DataFrameWriter[Row],DataFrameWriter[Row]] = {w=>w},
+                     metadataUpdateMethods: Option[List[String]] = None): Unit = {
+    require(
+      partition.nonEmpty || uri.isDefined,
+      "at least one partition must be specified"
+    )
+
+    require(
+      version.isDefined || owner.isDefined || uri.isDefined,
+      "version required to write global table partition"
+    )
+
+    val (detectedTableUri, detectedVersion, sqlTableName) = getTableInfo(
+      tableName=tableName,
+      version=version,
+      owner=owner,
+      uri=uri,
+      spark=spark,
+      adHocTablesDir=adHocTablesDir,
+      globalTablesDir=globalTablesDir
+    )
+
+    require(
+      detectedTableUri.isDefined,
+      s"table is not external: ${sqlTableName.getOrElse(tableName)}"
+    )
+
+    // maybe find partition string
+    val partitionString = if (partition.nonEmpty) {
+      Some(partition.map{p=>s"${p._1}=${p._2}"}.mkString("/"))
+    } else {
+      None
+    }
+
+    // build uri
+    val detectedUri: String = List(
+      Some(uri.getOrElse(detectedTableUri.get)),
+      partitionString
+    ).flatten.mkString("/")
+
+    if (sqlTableName.isEmpty && owner.isEmpty && uri.isEmpty) {
+      logger.warn(s"writing non-catalog global table partition: $detectedUri")
+    }
+
+    log(
+      "writePartition",
+      Map(
+        "detectedUri"->Some(detectedUri),
+        "detectedVersion"->detectedVersion,
+        "owner"->owner,
+        "partition"->partitionString,
+        "sqlTableName"->sqlTableName,
+        "tableName"->Some(tableName),
+        "uri"->uri,
+        "version"->version
+      )
+    )
+
+    extraWriteConfig(writeConfig(df.write)).save(detectedUri)
+
+    // update metadata on global tables
+    if (sqlTableName.isDefined) {
```

Comment:
In the future, we can use `ALTER TABLE ADD PARTITION` instead of `MSCK REPAIR TABLE` here.